dataset/src/test/scala/frameless/functions/NonAggregateFunctionsTests.scala

```diff
@@ -462,6 +461,72 @@ class NonAggregateFunctionsTests extends TypedDatasetSuite {
     check(forAll(prop[Int] _))
   }
 
+  test("inputFileName") {
+    val spark = session
+    import spark.implicits._
+
+    def prop[A : TypedEncoder](
+      toFile1: List[X1[A]],
+      toFile2: List[X1[A]],
+      inMem: List[X1[A]]
+    )(implicit x2Gen: Encoder[X2[A, String]], x3Gen: Encoder[X3[A, String, String]]) = {
+
+      val file1Path = "target/file1"
+      val file2Path = "target/file2"
+
+      val toFile1WithName = toFile1.map(x => X2(x.a, "file1"))
+      val toFile2WithName = toFile2.map(x => X2(x.a, "file2"))
+      val inMemWithName = inMem.map(x => X2(x.a, ""))
+
+      toFile1WithName.toDS().write.mode(SaveMode.Overwrite).parquet(file1Path)
+      toFile2WithName.toDS().write.mode(SaveMode.Overwrite).parquet(file2Path)
+
+      val readBackIn1 = spark.read.parquet(file1Path).as[X2[A, String]]
+      val readBackIn2 = spark.read.parquet(file2Path).as[X2[A, String]]
+
+      val ds1 = TypedDataset.create(readBackIn1)
+      val ds2 = TypedDataset.create(readBackIn2)
+      val ds3 = TypedDataset.create(inMemWithName)
+
+      val unioned = ds1.union(ds2).union(ds3)
+
+      val withFileName = unioned.withColumn[X3[A, String, String]](inputFileName[X2[A, String]]())
+        .collect()
+        .run()
+        .toVector
+
+      val grouped = withFileName.groupBy(_.b).mapValues(_.map(_.c).toSet)
+
+      grouped.foldLeft(passed) { (p, g) =>
+        p && secure { g._1 match {
+          case "" => g._2.head == "" //Empty string if didn't come from file
+          case f => g._2.forall(_.contains(f))
+        }}}
+    }
+
+    check(forAll(prop[String] _))
+  }
+
+  test("monotonic id") {
+    val spark = session
+    import spark.implicits._
+
+    def prop[A : TypedEncoder](xs: List[X1[A]])(implicit x2en: Encoder[X2[A, Long]]) = {
+      val ds = TypedDataset.create(xs)
+
+      val result = ds.withColumn[X2[A, Long]](monotonicallyIncreasingId[X1[A]]())
```

Comment:
Would this work if we omitted `X1[A]` and just type `monotonicallyIncreasingId()`?

Reply:
No, and it's terribly frustrating. Being that `monotonicallyIncreasingId` doesn't take a `TypedColumn[T, A]` like so many other functions do, there's no way that I know of to infer the `T` needed for the methods on `TypedDataset[T]` that require the `T`s to align, e.g. `select`, `withColumn` etc. 

Any ideas on that? Very open to suggestions here!

Reply:
if you help the compiler by specifically  adding `.apply` it can infer the type. I run into this few other times. For example, ` ds.withColumn[X2[A,Long]].apply(monotonicallyIncreasingId())` should work.

Reply:
The "problem" here is the existence of two overloaded methods withColumn. I am thinking maybe is a good idea to rename the withColumn that does replacement of existing column as withColumnReplaced. I tried it and with that the expression in question works without having to help the compiler with `.apply`.

Reply:
Agreed. I prefer renaming to `withColumnReplaced` over having to use types everywhere.